dipy/align/transforms.pyx

```diff
@@ -0,0 +1,955 @@
+#!python
+#cython: boundscheck=False
+#cython: wraparound=False
+#cython: cdivision=True
+
+import numpy as np
+cimport numpy as cnp
+cimport cython
+
+cdef extern from "dpy_math.h" nogil:
+    double cos(double)
+    double sin(double)
+    double log(double)
+
+cdef class Transform:
+    r""" Base class (contract) for all transforms for affine image registration.
+    Each transform must define the following (fast, nogil) methods:
+
+    1. _jacobian(theta, x, J): receives a parameter vector theta, a point in
+       x, and a matrix J with shape (dim, len(theta)). It must writes in J, the
+       Jacobian of the transform with parameters theta evaluated at x.
+
+    2. _get_identity_parameters(theta): receives a vector theta whose length is
+       the number of parameters of the transform and sets in theta the values
+       that define the identity transform.
+
+    3. _param_to_matrix(theta, T): receives a parameter vector theta, and a
+       matrix T of shape (dim + 1, dim + 1) and writes in T the matrix
+       representation of the transform with parameters theta
+
+    This base class defines the (slow, convenient) python wrappers for each
+    of the above functions, which also do parameter checking and raise
+    a ValueError in case the provided parameters are invalid.
+    """
+    cdef:
+        int number_of_parameters
+        int dim
+
+    def __cinit__(self):
+        r""" Default constructor
+        Sets transform dimension and number of parameter to invalid values (-1)
+        """
+        self.dim = -1
+        self.number_of_parameters = -1
+
+    cdef int _jacobian(self, double[:] theta, double[:] x, double[:,:] J)nogil:
+        return -1
+
+    cdef void _get_identity_parameters(self, double[:] theta) nogil:
+        return
+
+    cdef void _param_to_matrix(self, double[:] theta, double[:,:] T)nogil:
+        return
+
+    def jacobian(self, double[:] theta, double[:] x):
+        r""" Jacobian function of this transform
+
+        Parameters
+        ----------
+        theta : array, shape (n,)
+            vector containing the n parameters of this transform
+        x : array, shape (dim,)
+            vector containing the point where the Jacobian must be evaluated
+
+        Returns
+        -------
+        J : array, shape (dim, n)
+            Jacobian matrix of the transform with parameters theta at point x
+        """
+        n = len(theta)
+        if n != self.number_of_parameters:
+            raise ValueError("Invalid number of parameters: %d"%(n,))
+        m = len(x)
```

Comment:
Right, now using .shape[0]. I didn't worry about this one, though, because _jacobian(...) will be called instead during optimization.